mempool/src/pool.rs

```diff
@@ -0,0 +1,2787 @@
+            self.store.drop_tx_and_descendants(removed.tx.get_id());
+        }
+        removed_fees
+    }
+}
+
+trait SpendsUnconfirmed<C, T, M>
+where
+    C: ChainState,
+    T: GetTime,
+    M: GetMemoryUsage,
+{
+    fn spends_unconfirmed(&self, mempool: &MempoolImpl<C, T, M>) -> bool;
+}
+
+impl<C, T, M> SpendsUnconfirmed<C, T, M> for TxInput
+where
+    C: ChainState,
+    T: GetTime,
+    M: GetMemoryUsage,
+{
+    fn spends_unconfirmed(&self, mempool: &MempoolImpl<C, T, M>) -> bool {
+        mempool.contains_transaction(
+            self.get_outpoint().get_tx_id().get_tx_id().expect("Not coinbase"),
+        )
+    }
+}
+
+#[derive(Clone)]
+struct SystemClock;
+impl GetTime for SystemClock {
+    fn get_time(&self) -> Time {
+        common::primitives::time::get()
+    }
+}
+
+#[derive(Clone)]
+struct SystemUsageEstimator;
+impl GetMemoryUsage for SystemUsageEstimator {
+    fn get_memory_usage(&self) -> MemoryUsage {
+        //TODO implement real usage estimation here
+        0
+    }
+}
+
+impl<C, T, M> Mempool<C, T, M> for MempoolImpl<C, T, M>
+where
+    C: ChainState,
+    T: GetTime,
+    M: GetMemoryUsage,
+{
+    fn create(chain_state: C, clock: T, memory_usage_estimator: M) -> Self {
+        Self {
+            store: MempoolStore::new(),
+            chain_state,
+            max_size: MAX_MEMPOOL_SIZE_BYTES,
+            max_tx_age: DEFAULT_MEMPOOL_EXPIRY,
+            rolling_fee_rate: Cell::new(RollingFeeRate::new(clock.get_time())),
+            clock,
+            memory_usage_estimator,
+        }
+    }
+
+    fn new_tip_set(&mut self, chain_state: C) {
+        self.chain_state = chain_state;
+        self.rolling_fee_rate.set({
+            let mut rolling_fee_rate = self.rolling_fee_rate.get();
+            // TODO Not sure we should set the flag to true when a block is disconnected/during a
+            // reorg
+            rolling_fee_rate.block_since_last_rolling_fee_bump = true;
+            rolling_fee_rate
+        })
+    }
+    //
+
+    fn add_transaction(&mut self, tx: Transaction) -> Result<(), Error> {
+        let conflicts = self.validate_transaction(&tx)?;
+        self.store.drop_conflicts(conflicts);
+        self.finalize_tx(tx)?;
+        Ok(())
+    }
+
+    fn get_all(&self) -> Vec<&Transaction> {
+        self.store
+            .txs_by_descendant_score
+            .values()
+            .flatten()
+            .map(|id| &self.store.get_entry(id).expect("entry").tx)
+            .collect()
+    }
+
+    fn contains_transaction(&self, tx_id: &Id<Transaction>) -> bool {
+        self.store.txs_by_id.contains_key(&tx_id.get())
+    }
+
+    // TODO Consider returning an error
+    fn drop_transaction(&mut self, tx_id: &Id<Transaction>) {
+        self.store.remove_tx(tx_id);
+    }
+}
+
+fn has_duplicate_entry<T>(iter: T) -> bool
+where
+    T: IntoIterator,
+    T::Item: Ord,
+{
+    let mut uniq = BTreeSet::new();
+    iter.into_iter().any(move |x| !uniq.insert(x))
+}
+
+#[cfg(test)]
```

Comment:
Please move the tests to another sub-crate. This file is really huge.

Also consider splitting this further. Again, you don't have to do this right now. But it has to be done. The file is huge!

Reply:
I tried moving the tests to another file in the past, but ran into issues because as far as I could see in Rust tests that access private functions/members (and we have some such tests, `tx_mempool_entry` to give just one example) need to be in the same file as the module. I'll take another look to see if something can be done, probably some of the test can be moved.

Reply:
See how Carla did it in Utxo for private stuff.

Reply:
At the very least you should be able to take the test module and move it to `src/pool/test.rs`. Here, there would only be `#[cfg(test)] mod test;`. That way the module structure and therefore visibility rules remain the same as are now.

Reply:
Done.

Reply:
I think you forgot to check in the actual file?